### DISCUSSION TITLE
Some thoughts on AnimatedTexture

### DISCUSSION THREAD

**Discussion by leewyatt** [AlmasB/FXGL#1173] [Ideas]
Hi ~
    Very happy, the new version of FXGL supports AnimatedTexture reverse play ; I still have some immature ideas

**1.** 
Many times, I think AnimatedTexture has a state, the code will be more concise, such as knowing whether it is currently playing;
`AnimatedTexture.getStatus()`
I have to add a variable **stopFlag** to store the current state
![image](https://user-images.githubusercontent.com/75261429/164952013-4cc80b2c-62ca-48a8-b4ad-cd6d01fc8e8d.png)

**2.**  
![image](https://user-images.githubusercontent.com/75261429/164954190-b462578c-3e4a-4db8-9d62-e9e8aed06ced.png)

For example, bucket zombies have multiple walking states;
a. Bucket complete img1
b. The bucket is damaged img2
c. The bucket is seriously damaged img3
d. The bucket is shot down img4 
Suppose when the bucket zombie (img1) plays to the 12th frame, the bucket is damaged, I need to start playing from the 13th frame of img2,
When playing the 18th frame, the bucket is seriously damaged, then I need to start from the 19th frame of img3 to play;

To start playing from the specified frame, we can create a new AnimationChannel specifying the start and end frames;
Now there is no way to get the index of the currently playing frame;

**3.**
![image](https://user-images.githubusercontent.com/75261429/164954002-7db25573-d157-49e4-96c0-382f4b71362a.png)
![image](https://user-images.githubusercontent.com/75261429/164954008-1a2e243b-8078-4682-b5c3-23fc93f046db.png)

Such as the combination of buckets and zombies to form bucket zombies;
But every frame, the bucket will rotate at a certain angle, and the coordinate position will be different.
The following two solutions I can think of at the moment;
1). Process it in Photoshop; if I do this, the number of image files will become a lot, and the scalability seems to be bad, for example, the traffic cone zombie is formed by adding traffic cones to zombies;
2). Write a JSON file, which defines each frame, the image of the zombie, the rotation angle and coordinate position of the bucket. Of course, this requires some methods provided by FXGL (AnmaitonChannel) to support.

**4**
![image](https://user-images.githubusercontent.com/75261429/164973792-763c7724-bf4c-4cc2-a837-43ae16316f94.png)
The position of the weapon in each frame is different, that is, the Hitbox of each frame of the picture is also different, and it seems that json is also needed to describe.

I saw on github that in the json files of some games, dozens of properties such as "ScaleX", "ScaleY", "Position", "opacity", "visiable", etc. are defined;

---------------------------------------------------------------------------------------------------------
The above things seem to be a lot of work;

Although I can do it by parsing JSON and using javafx's ImageView and animation, I am studying economics and have no rich game development experience, so I can't seem to design a general, high-performance, high-quality API;

If FXGL can support the above functions, then it is more perfect in 2D games;

FXGL is very interesting~ 

Thank you~

**Comment by leewyatt** [comment]
I don't know if using **spine** technology can solve the above problems

**Comment by AlmasB** [comment]
Thanks for these ideas, I've linked them against https://github.com/AlmasB/FXGL/issues/870 so they can be considered as a whole

**Comment by AlmasB** [comment]
@anivie thanks, for referring me to this discussion. Whilst this is not on my radar at the moment, I can take a brief look if you can provide an example, including:

1) the name and link to the software that is used to produce metadata files (e.g. json here) 
2) a sample .json file with file format specification (e.g. which fields are required and which are optional)
3) appropriate sprite sheet to go with the .json file
